UnifiedPDF: Show find indicator for the current match

hortont424 opened the pull request:
#### 15ec60645616ffbb888ccab283636ce395d39be2
<pre>
UnifiedPDF: Show find indicator for the current match
<a href="https://bugs.webkit.org/show_bug.cgi?id=269559">https://bugs.webkit.org/show_bug.cgi?id=269559</a>
<a href="https://rdar.apple.com/119632525">rdar://119632525</a>

Reviewed by Simon Fraser.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::textIndicatorForSelection):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::textIndicatorForSelection):
Allow PDF plugins to vend text indicators for the current selection, for
FindController&apos;s use (and hopefully other things later).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
Make it possible to paint only the foreground of the PDF.

(WebKit::UnifiedPDFPlugin::convertFromDocumentToContents const):
Add a rect version of document-&gt;contents and clean up the point version
slightly, though with no behavior change.

(WebKit::UnifiedPDFPlugin::textIndicatorForSelection):
Generate a TextIndicatorData manually (because TextIndicator only knows
how to deal in DOM ranges) for the current PDF selection.

(WebKit::UnifiedPDFPlugin::showDefinitionForAttributedString):
Leave a comment about potential future code sharing.

* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::updateFindUIAfterPageScroll):
(WebKit::FindController::updateFindIndicator):
Allow main frame PDF plugins to also desire and install find indicators.

(WebKit::FindController::indicateFindMatch):
(WebKit::FindController::showFindIndicatorInSelection):
(WebKit::FindController::deviceScaleFactorDidChange):
(WebKit::FindController::redraw):
(WebKit::FindController::didScrollAffectingFindIndicatorPosition):
Stop passing the frame into updateFindIndicator, since it can work it out
itself, and that allows us to put the plugin deferral code in one place,
like it is in the rest of FindController.

* Source/WebKit/WebProcess/WebPage/FindController.h:

Canonical link: <a href="https://commits.webkit.org/274875@main">https://commits.webkit.org/274875@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83e38f1cb8919fe3a41226765bc888308f73c168

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40239 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42617 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42784 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22178 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16580 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40813 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16209 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44062 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36490 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12342 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9036 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -54,6 +55,7 @@
 #include <WebCore/EditorClient.h>
 #include <WebCore/FilterOperations.h>
 #include <WebCore/FloatPoint.h>
+#include <WebCore/FrameSnapshotting.h>
```
didn't end up needing this

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -592,11 +594,14 @@ static String mutationObserverNotificationString()
     if (m_size.isEmpty() || documentSize().isEmpty())
         return;
 
+    bool shouldPaintBackground = behavior == PaintingBehavior::All;
```
Not sure if we need to bother making this an optionset already, but we could.
Also not sure how this is going to play with smfr's patch.

aprotyas replied:
oh I didn't realize there are other permutations? Something similar in smfr's patch?

hortont424 replied:
Likely we will need selection-but-no-background in the near future

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -151,6 +152,7 @@ class PDFPluginBase : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<PDF
     virtual bool findString(const String& target, WebCore::FindOptions, unsigned maxMatchCount) = 0;
     virtual Vector<WebCore::FloatRect> rectsForTextMatchesInRect(const WebCore::IntRect&) const { return { }; }
     virtual bool drawsFindOverlay() const = 0;
+    virtual RefPtr<WebCore::TextIndicator> textIndicatorForSelection(OptionSet<WebCore::TextIndicatorOption>, WebCore::TextIndicatorPresentationTransition) { return { }; }
```
Yay, perfect, thanks for this.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2495,6 +2545,8 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
     dictionaryPopupInfo.origin = rectInRootView.location();
     dictionaryPopupInfo.platformData.attributedString = WebCore::AttributedString::fromNSAttributedString(string.get());
 
+    // FIXME: Consider merging with textIndicatorForSelection in order to get
+    // a PDF snapshot instead of an attributed string replica.
```
Is a PDF snapshot more desirable?

aprotyas replied:
Certainly will merge the two soon™️

hortont424 replied:
Yes, because it looks exactly like the PDF! No slight jump/alignment issues. We learned this lesson many years ago with web content.

The one thing that's still better about web content is that we can do a selection-only paint so you don't get irrelevant rendering in your snapshot. But this is still progress.

aprotyas commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -492,31 +509,22 @@ bool FindController::shouldHideFindIndicatorOnScroll() const
 {
-    Ref selectedFrame = CheckedRef(m_webPage->corePage()->focusController())->focusedOrMainFrame();
-    updateFindIndicator(selectedFrame, false);
+    updateFindIndicator(false);
```
Nit: The `updateFindIndicator` args could use enum class-ification

aprotyas reviewed (commented):
LGTM

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2448,6 +2462,42 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
 
+RefPtr<TextIndicator> UnifiedPDFPlugin::textIndicatorForSelection(OptionSet<WebCore::TextIndicatorOption> options, WebCore::TextIndicatorPresentationTransition transition)
+{
+    if (auto selectionBounds = selectionBoundsForFirstPageInDocumentSpace(m_currentSelection)) {
```
Would be better to early return so the rest of the function is not indented.

smfr replied:
I'm confused by "selectionBoundsForFirstPageInDocumentSpace". First page of the document, or first page of the selection? Why is it OK to only consider the first page?

hortont424 replied:
Find highlights spanning multiple pages will just highlight the beginning of the match, like their Lookup and Define siblings. Room for improvement.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2448,6 +2462,42 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+        auto& context = buffer->context();
+        context.translate(-rectInContentsCoordinates.location());
+        context.scale(m_scaleFactor);
+        context.translate(centeringOffset());
```
I'm a bit confused by this. Would be nicer of when `paintPDFContent()` is called the context is always set up with the coordinate system that matches `m_contentsLayer` painting.

hortont424 replied:
That's what I'm doing, right? (but also shifting it so the origin is at the rect we're drawing)

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2448,6 +2462,42 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+        context.translate(centeringOffset());
+        paintPDFContent(context, rectInContentsCoordinates, PaintingBehavior::PageContentsOnly);
+
+        TextIndicatorData data;
+        data.contentImage = BitmapImage::create(ImageBuffer::sinkIntoNativeImage(WTFMove(buffer)));
+        data.contentImageScaleFactor = deviceScaleFactor;
+        data.contentImageWithoutSelection = data.contentImage;
+        data.contentImageWithoutSelectionRectInRootViewCoordinates = rectInRootViewCoordinates;
+        data.selectionRectInRootViewCoordinates = rectInRootViewCoordinates;
+        data.textBoundingRectInRootViewCoordinates = rectInRootViewCoordinates;
+        data.textRectsInBoundingRectCoordinates = { FloatRect { 0, 0, static_cast<float>(rectInRootViewCoordinates.width()), static_cast<float>(rectInRootViewCoordinates.height()) } };
+        data.presentationTransition = transition;
+        data.options = options;
```
Is this a pretty common set of fields that are set together for text indicators? If so maybe we could do some TextIndicator cleanup late to help with this case or we could just create a helper for the `UnifiedPDFPlugin``

hortont424 replied:
This is the helper. There are some improvements we can make here (vending one with the selection but still no background on `contentImage` and without the selection on `contentImageWithoutSelection`, etc.) to improve transitions, but it's a start.

sammygill commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -228,7 +228,12 @@ void FindController::updateFindUIAfterPageScroll(bool found, const String& strin
     }
 
-    if (found && (!options.contains(FindOptions::ShowFindIndicator) || !selectedFrame || !updateFindIndicator(*selectedFrame, shouldShowOverlay)))
+    bool wantsFindIndicator = found && options.contains(FindOptions::ShowFindIndicator);
+    bool canShowFindIndicator = selectedFrame;
```
Is it the frame that wants the find indicator/can show it or something else?

hortont424 replied:
It's the frame containing the DOM selection (irrelevant in our case since we don't have one).

hortont424 committed:
UnifiedPDF: Show find indicator for the current match
https://bugs.webkit.org/show_bug.cgi?id=269559
rdar://119632525

Reviewed by Simon Fraser.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::textIndicatorForSelection):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::textIndicatorForSelection):
Allow PDF plugins to vend text indicators for the current selection, for
FindController's use (and hopefully other things later).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
Make it possible to paint only the foreground of the PDF.

(WebKit::UnifiedPDFPlugin::convertFromDocumentToContents const):
Add a rect version of document->contents and clean up the point version
slightly, though with no behavior change.

(WebKit::UnifiedPDFPlugin::textIndicatorForSelection):
Generate a TextIndicatorData manually (because TextIndicator only knows
how to deal in DOM ranges) for the current PDF selection.

(WebKit::UnifiedPDFPlugin::showDefinitionForAttributedString):
Leave a comment about potential future code sharing.

* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::updateFindUIAfterPageScroll):
(WebKit::FindController::updateFindIndicator):
Allow main frame PDF plugins to also desire and install find indicators.

(WebKit::FindController::indicateFindMatch):
(WebKit::FindController::showFindIndicatorInSelection):
(WebKit::FindController::deviceScaleFactorDidChange):
(WebKit::FindController::redraw):
(WebKit::FindController::didScrollAffectingFindIndicatorPosition):
Stop passing the frame into updateFindIndicator, since it can work it out
itself, and that allows us to put the plugin deferral code in one place,
like it is in the rest of FindController.

* Source/WebKit/WebProcess/WebPage/FindController.h:

Canonical link: https://commits.webkit.org/274875@main
